quic/s2n-quic-core/src/datagram/default.rs

```diff
@@ -3,16 +3,161 @@
 
 // s2n-quic's default implementation of the datagram component
 
-use crate::datagram::{Packet, Sender};
+use core::task::{Context, Poll, Waker};
+
+use crate::datagram::{ConnectionInfo, Packet, Sender};
 use alloc::collections::VecDeque;
 use bytes::Bytes;
 
+#[derive(Debug)]
 pub struct DefaultSender {
-    queue: VecDeque<Datagram>,
+    pub queue: VecDeque<Datagram>,
+    capacity: usize,
+    min_packet_space: usize,
+    max_packet_space: usize,
+    smoothed_packet_size: f64,
+    waker: Option<Waker>,
+    max_datagram_payload: u64,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Datagram {
+    pub data: Bytes,
+}
+
+#[non_exhaustive]
+#[derive(Debug, PartialEq)]
+pub enum SendDatagramError {
+    QueueAtCapacity,
+    DataLengthExceedsPeerLimits,
 }
 
-struct Datagram {
-    data: Bytes,
+impl DefaultSender {
+    /// Creates a builder for the default datagram sender
+    pub fn builder() -> Builder {
+        Builder::default()
+    }
+
+    /// Enqueues a datagram for sending it towards the peer.
+    ///
+    /// # Return value
+    ///
+    /// The function returns:
+    ///
+    /// - `Poll::Pending` if the datagram's send buffer capacity is currently exhausted
+    ///   and the datagram was not added to the queue. In this case, the caller should
+    ///   retry sending after the [`Waker`](core::task::Waker) on the provided
+    ///   [`Context`](core::task::Context) is notified.
+    /// - `Poll::Ready(Ok(()))` if the datagram was enqueued for sending.
+    /// - `Poll::Ready(Err(SendDatagramError))` if an error occurred while trying
+    ///   to send the datagram
+    pub fn poll_send_datagram(
+        &mut self,
+        data: bytes::Bytes,
+        cx: &mut Context,
+    ) -> Poll<Result<(), SendDatagramError>> {
+        if data.len() as u64 > self.max_datagram_payload {
+            return Poll::Ready(Err(SendDatagramError::DataLengthExceedsPeerLimits));
+        }
+
+        if self.queue.len() == self.capacity {
+            self.waker = Some(cx.waker().clone());
+            return Poll::Pending;
+        }
+
+        let datagram = Datagram { data };
+        self.queue.push_back(datagram);
+        Poll::Ready(Ok(()))
+    }
+
+    /// Adds datagrams on the queue to be sent
+    ///
+    /// If the datagram queue is at capacity the oldest datagram will be popped
+    /// off the queue and returned to make space for the newest datagram.
+    pub fn send_datagram(
```

Comment:
I would call this something like `send_datagram_forced` and the other one `send_datagram`